src/api/libcellml/component.h

```diff
@@ -0,0 +1,63 @@
+#ifndef LIBCELLML_LIBCELLML_COMPONENT_H
+#define LIBCELLML_LIBCELLML_COMPONENT_H
+
+#include <string>
+#include <memory>
+
+#include "libcellml/libcellml_export.h"
+#include "libcellml/nameable.h"
+
+namespace libcellml {
+
+/**
+ * @brief The Model class.
+ * The Model class is for representing a CellML Model.
+ */
+class LIBCELLML_EXPORT Component: public Nameable
+{
+public:
+    /**
+     * @brief Component Constructor.
+     * Default Component constructor.
+     */
+    Component();
+    /**
+     * @brief Component Destructor.
+     * Default Component destructor.
+     */
+    ~Component();
+
+    /**
+     * @brief serialise the Model.
+     * This method serialises the Model to a std:string representation in the
+     * desired format.
+     * @sa deserialise
+     * @param format The format to serialise the Model to.
+     * @return std::string representation of the Model in the given format.
+     */
+    std::string serialise(CELLML_FORMATS format) const;
+
+    Component(Component&& rhs); /**< move constructor */
+    Component& operator=(Component&& rhs); /**< move assignment */
+
+    /**
+     * @brief add a child component to this component.
+     * Add a child component to this component.
+     * @param c the component to add.
+     */
+    void addComponent(Component &c);
+
+private:
+
+    Component(const Component&) = delete; /**< non-copyable */
+    Component& operator=(const Component&) = delete; /**< non-copyable */
+
+    struct ComponentImpl; /**< Forward declaration for pImpl idiom. */
+    std::unique_ptr<ComponentImpl> mPimpl; /**< Private member to implementation pointer */
```

Comment:
If you're going to store a list of child Component objects I'd expect it to be an STL container of shared_ptr, leading to the conclusion that all objects are on the heap (never on the stack) with lifetimes managed by shared_ptr references. [Unlike the intrusive reference counting we use in Zinc, my understanding is that shared_ptr is implemented as a linked list, requiring disciplined use of shared_ptr for all methods that affect object lifetime: create(), addComponent() etc. For using other methods it is perfectly fine or even preferred to use a reference to the actual object (i.e. obtain it from the shared_ptr).]